crates/trie/src/state.rs

```diff
@@ -42,26 +50,74 @@ impl HashedPostState {
-            if let Some(account) = &account.info {
-                hashed_state.insert_account(hashed_address, into_reth_acc(account.clone()))
-            } else {
-                hashed_state.insert_destroyed_account(hashed_address);
-            }
+            this.insert_account(hashed_address, account.info.clone().map(into_reth_acc));
 
             // insert storage.
             let mut hashed_storage = HashedStorage::new(account.status.was_destroyed());
 
             for (key, value) in account.storage.iter() {
                 let hashed_key = keccak256(B256::new(key.to_be_bytes()));
-                hashed_storage.insert_storage(hashed_key, value.present_value);
+                hashed_storage.insert_slot(hashed_key, value.present_value);
+            }
+            this.insert_hashed_storage(hashed_address, hashed_storage)
+        }
+        this.sorted()
+    }
+
+    /// Initialize [HashedPostState] from revert range.
+    /// Iterate over state reverts in the specified block range and
+    /// apply them to hashed state in reverse.
+    ///
+    /// NOTE: In order to have the resulting [HashedPostState] be a correct
+    /// overlay of the plain state, the end of the range must be the current tip.
+    pub fn from_revert_range<TX: DbTx>(
+        tx: TX,
+        range: RangeInclusive<BlockNumber>,
+    ) -> Result<Self, DatabaseError> {
+        let mut state =
+            HashMap::<Address, (Option<Option<Account>>, HashMap<B256, U256>)>::default();
+
+        // Iterate over account changesets in reverse.
+        let account_changesets = tx
+            .cursor_read::<tables::AccountChangeSet>()?
+            .walk_range(range.clone())?
+            .collect::<Result<Vec<_>, _>>()?;
+        for (_, account_before) in account_changesets.into_iter().rev() {
+            let AccountBeforeTx { address, info } = account_before;
+            state.entry(address).or_default().0 = Some(info);
+        }
+
+        // Iterate over storage changesets in reverse.
+        let storage_range = BlockNumberAddress::range(range);
+        let storage_changesets = tx
+            .cursor_read::<tables::StorageChangeSet>()?
+            .walk_range(storage_range)?
+            .collect::<Result<Vec<_>, _>>()?;
+        for (block_number_and_address, storage) in storage_changesets.into_iter().rev() {
+            let BlockNumberAddress((_, address)) = block_number_and_address;
+            state.entry(address).or_default().1.insert(storage.key, storage.value);
+        }
+
+        let mut this = Self::default();
+        for (address, (maybe_account_change, storage)) in state {
+            let hashed_address = keccak256(address);
+
+            if let Some(account_change) = maybe_account_change {
+                this.insert_account(hashed_address, account_change);
             }
-            hashed_state.insert_hashed_storage(hashed_address, hashed_storage)
+
+            let mut hashed_storage = HashedStorage::new(false);
```

Comment:
is always-false here right? does it not need to depend on whether `account_change` is `Some` or `None`?

Reply:
the wiped flag only dictates whether previous storage entries should be looked up in db or not. for reverts it's a noop since all wiped changes had been written as storage reverts